### ISSUE TITLE
Bug with ssripat_scope?

### ISSUE THREAD

**Issue by CohenCyril** [coq/coq#10531]
#### Description of the problem

Bug in #6705, the scope ssripat_scope seems to be used after a clear `{}` in a `rewrite`. See

```coq
From Coq Require Import ssreflect.

Notation "'!' rule" := (ltac:(do !rewrite rule/=))
   (at level 0, rule at level 0) : ssripat_scope.

Goal forall x : bool, false = true -> true = false.
Proof.
move=> b x.
Fail rewrite {b} !x.
rewrite !x.  (*and so does rewrite {b} /= x*)
Abort.
````

#### Coq Version

Coq 8.10+beta1 (June 2019)

**Comment by gares** [comment]
I think that it is not a bug. Even if the notation is in a specific scope, what you are changing is the grammar for terms. A rewrite rule is parsed as a term and here the full grammar of a rewrite "argument" is ambiguous since you don't know if `!` is part of the term or is the multiplier.

**Comment by gares** [comment]
Now, there is still a very fishy thing, since the ambiguity seems to show up only after `{stuff}`.
https://github.com/coq/coq/blob/a3de6d9788354b011c8f1111163dab1dff81379a/plugins/ssr/ssrparser.mlg#L2352-L2372
The only difference seems to be that we parse an `ssrrule` (L2363) instead of a `ssrrule_ne` that time (that is we allow `{stuff}` to be the last `rewrite` argument, which makes sense but can be expressed differently I suppose).

The two grammar entries are defined one in terms of the other, so I've no clue why the behavior is different.
https://github.com/coq/coq/blob/a3de6d9788354b011c8f1111163dab1dff81379a/plugins/ssr/ssrparser.mlg#L2309-L2324

Let's put @ggonthier in the loop.

Bottom line, the issue has nothing to to with the ssripat_scope I'm afraid.

**Comment by ggonthier** [comment]
This behaviour results from the quirks of the camlp5 parsing engine. In retrospect, defining `constr` syntax for '`!`' at level < 10 clearly interferes with the `rewrite` arguments syntax, so it's unsurprising you get  weird results.
   The relevant `rewrite` argument syntax fragment is [`!`][`{`_idents_`}`][_term_], where each of the bracketed element are optional (but not all), with the proviso that _term_ must be present when there is a `!`. Observe that the `!` _precedes_ the clear/select item `{`_idents_`}`. The grammar tries to parse this as a whole because all three parts need to be interpreted together, and to try to give better error messages.
   In @CohenCyril's second line the `!` is parsed by the rewrite argument rule (as an `ssrmult_ne`), so the new user parsing rule is not used. In the first line, a `constr` is expected after the `}`, so the notation _is_ used, but since presumably `ssripat_scope` is not open at this point, you get an interpretation error. Without the notation, `!r` would not have parsed as a `constr`, so `{b}` would have parsed as a separate argument, and then `!r` would have parsed as a second argument to `rewrite`, as in the second line.
  There are a few things askew here:
  - `ssripat_scope` notation interfere with the general `constr` grammar because it is a (normal) interpretation scope, rather than a parsing scope tied to a separate nonterminal (which Coq 8.10 now supports).
  - `ssrrule` does not guard against separate arguments that could be partially parsed as terms.
In this case the conflict was created by new user syntax, but it already exists for numeric multipliers and clear switches, e.g.,
```
rewrite {n} 2!addnC.
rewrite {n} {m} /=.
```
both give off weird diagnostics because they try to parse and use `2` and `{m}` as rewrite rules, respectively. We could use lookahead to either allow these to work properly (or give more meaningful errors, if you want to disallow isolated clears, though I'm not inclined to).
